src/checksum/arm64/crc32-arm64.c

```diff
@@ -59,30 +59,45 @@ static uint32_t
 __wt_checksum_with_seed_hw(uint32_t seed, const void *chunk, size_t len)
 {
     uint32_t crc;
-    size_t nqwords;
-    const uint8_t *p;
     const uint64_t *p64;
+    const uint32_t *p32;
+    const uint16_t *p16;
+    const uint8_t *p8;
 
     crc = ~seed;
 
     /* Checksum one byte at a time to the first 4B boundary. */
-    for (p = chunk; ((uintptr_t)p & (sizeof(uint32_t) - 1)) != 0 && len > 0; ++p, --len) {
-        CRC32CB(crc, *p);
+    for (p8 = chunk; ((uintptr_t)p8 & (sizeof(uint32_t) - 1)) != 0 && len > 0; ++p8, --len)
+        CRC32CB(crc, *p8);
+
+    p64 = (const uint64_t *)p8;
+    /* Checksum in 16B chunks. */
+    for (; len >= 16; p64 += 2, len -= 16) {
+        CRC32CX(crc, *p64);
+        CRC32CX(crc, *(p64 + 1));
```

Comment:
Would you mind checking the asm and verifying that this generates a single LDP (load pair) instruction rather than an LDR (load register)?

Reply:
```
   0x00000000000ea7c8 <+40>:	ldr	x3, [x1]
   0x00000000000ea7cc <+44>:	crc32cx	w0, w0, x3
   0x00000000000ea7d0 <+48>:	ldr	x3, [x1, #8]
   0x00000000000ea7d4 <+52>:	crc32cx	w0, w0, x3
```

@RedBeard0531  I looked at the assembly code. It is still generating ldr instructions. Any idea why? We use mongodbtoolchain v4 to build.

Reply:
Here's the result of the release build:

```
   0x00000000000d3be8 <+72>:	ldr	x3, [x1]
   0x00000000000d3bec <+76>:	add	x1, x1, #0x10
   0x00000000000d3bf0 <+80>:	crc32cx	w0, w0, x3
   0x00000000000d3bf4 <+84>:	ldur	x3, [x1, #-8]
   0x00000000000d3bf8 <+88>:	crc32cx	w0, w0, x3
```

Reply:
I changed the code to:

```
    for (; len >= 16; p64 += 2, len -= 16) {
        a = *p64;
        b = *(p64 + 1);
        CRC32CX(crc, a);
        CRC32CX(crc, b);
    }
```
It is still the same.

Reply:
I think the main issue is I'm missing the compiler flag mcpu=neoverse-n1. Will try tomorrow.

Reply:
@RedBeard0531 I'm still not able to generate the ldp instructions with `mcpu=neoverse-n1` on graviton2. Can you give more details about your testing environment? Also in your code, I see:

```
loadAt<uint64_t>(p)
```
Is this your own implementation? Will it make a difference?

Reply:
You can see the full class with the `loadAt<T>` method [here](https://github.com/10gen/mongo/blob/4cb0df8d7f36b3091d6d653c8c9f7242f29e92e4/src/mongo/db/storage/wiredtiger/joplog_record_store.cpp#L72-L145).

I popped it into [godbolt](https://www.godbolt.org/z/9hezGr3KP), and it looks like the ability to convert 2 loads to an `ldp` was added in gcc 12, but we are still using gcc 11.3 😞. I could have sworn that I saw an `ldp` when I was looking at the `perf annotate` disasm output after building on our toolchain, but I guess I must have only looked at godbolt with the ARM64 gcc trunk compiler (that was were I wrote that code initially to make sure it was generating good asm as I went). Either way I was benchmarking with gcc 11.3, so I must have benchmarked with 2 `ldr`s, which means that doing the 2 loads per loop still seems to be a win. The CPU probably executes them in parallel anyway, since it has 2 load EUs.

So either we should just accept that output from our current toolchain and let it get converted to use `ldp` when we upgrade, or we write that part of the loop in inline asm. Since we are already using asm for the `CRC32CX` macro, maybe that isn't so crazy? That would also let us use the post-increment form of `ldp` which gcc doesn't seem to be able to do automatically (strangely it can when using `ldr`), and save one instruction in that loop. I think you want something like the following, which will generate the nice output from left-most column [here](https://www.godbolt.org/z/TPTh5M1n9).

```cpp
       if (bytes >= 16) {
            auto end = p + (bytes & ~0b1111);
            uint64_t a, b;
            asm (
                "1: \n\t" // loop label
                "ldp %[a], %[b], [%[p]], #16 \n\t" // post-increment
                "crc32cx %w[reg], %w[reg], %[a] \n\t"
                "crc32cx %w[reg], %w[reg], %[b] \n\t"
                "cmp %[p], %[end] \n\t"
                "b.ne 1b"
                : [p] "+&r" (p)
                , [reg] "+&r" (reg)
                , [a] "=&r" (a)
                , [b] "=&r" (b)
                : [end] "r" (end)
                : "cc" // clobbers condition codes
            );
        }
```

Reply:
Nevermind, that asm is actually slower. I don't understand why, but somehow, the looping logic generated by the compiler is faster. It _isn't_ because of `ldp` vs `ldr` and it isn't because this makes the later branches depend on loop state. I can get identical perf to the current codegen by just putting the `ldp` and `crc32cx` instructions in the `asm` block and putting them in the loop with a `bytes -= 16;`. So I guess 1 `ldp` vs 2 `ldr`s doesn't actually help, at least in this case. So it isn't worth using more inline asm.

That said, it seems like the asm you are getting is very different from what I'm seeing. At the very least, yours isn't using the post-increment and is interleaving the loads and crc, rather than doing both of each type. No idea if that matters for perf though. As a quick sanity check could you try running this very simplistic test:

```cpp
char zeros[1024];
volatile uint32_t output;
int main() {
        for (int i= 0; i<18*1024*1024; i++) {
                output = hashBytes(zeros, 1024);
                asm volatile("":::"memory"); // Prevent the compiler from assuming that zeros is unmodified.
        }
}
```
I'm seeing that take 0.976 seconds to process 18GB, 1KB at a time. If you are seeing the same perf, then the details of what the compiler is emitting don't matter. If you aren't then we can dig into why not, but lets not waste time if it doesn't matter.